product-mini/platforms/posix/main.c

```diff
@@ -934,38 +939,52 @@ main(int argc, char *argv[])
         native_lib_list, native_lib_count, native_lib_loaded_list);
 #endif
 
-    /* load WASM byte buffer from WASM bin file */
-    if (!(wasm_file_buf =
-              (uint8 *)bh_read_file_to_buffer(wasm_file, &wasm_file_size)))
+    fd = open(wasm_file, O_RDONLY, 0);
+
+    if (fd < 0) {
+        printf("Open wasm app file [%s] failed.\n", wasm_file);
         goto fail1;
+    }
 
-#if WASM_ENABLE_AOT != 0
-    if (wasm_runtime_is_xip_file(wasm_file_buf, wasm_file_size)) {
+    if (fstat(fd, &wasm_file_info) < 0) {
+        printf("Get wasm app file info [%s] failed.\n", wasm_file);
+        goto fail1;
+    }
+
+    /* mmap the wasm file */
+    if ((wasm_file_buf = mmap(NULL, wasm_file_info.st_size,
+                              PROT_READ | PROT_EXEC, MAP_PRIVATE, fd, 0))
```

Comment:
doesn't this break platforms like this?
https://github.com/bytecodealliance/wasm-micro-runtime/blob/797e534cece680fae65734560298ac11fc550a0f/core/shared/platform/nuttx/nuttx_platform.c#L94-L112

anyway, i suspect you should not use mmap directly.
instead, you somehow need to extend the platform api to perform file-mmap where supported (and fallback otherwise), i guess.

Reply:
OK, Seems extended the os_mmap to supoort map file from given fd is better.
What's your idea? @wenyongh

Reply:
Yes, normally we should not use mmap directly, instead we had better extend os_mmap and add os_open/os_fstat if needed.

Fortunately, the branch dev/wasi-libc-windows has added API os_open_preopendir and os_fstat in platform_api_extension.h and implemented them in posix platforms and windows platform. See:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/dev/wasi-libc-windows/core/shared/platform/include/platform_api_extension.h#L1210-L1234

And now we are merging branch dev/wasi-libc-windows into main, the PR is under review:
https://github.com/bytecodealliance/wasm-micro-runtime/pull/2740

I think we can wait until the PR is merged and then extend the os_mmap.